packages/devtools_app/lib/src/screens/memory/shared/heap/heap_analyzer.dart

```diff
@@ -64,5 +117,29 @@ bool _canRetain(String klass, String library) {
   // or detect weak references automatically, without hard coding
   // class names.
   assert(false, 'Unexpected library for $klass: $library.');
+  return false;
+}
+
+/// Verifies heap integrity rules.
+///
+/// 1. Nullness of 'retainedSize' and 'retainer' should be equal.
+///
+/// 2. Root's 'retainedSize' should be sum of shallow sizes of all reachable
+/// objects.
+bool _verifyHeapIntegrity(AdaptedHeap heap) {
+  var totalReachableSize = 0;
+
+  for (var object in heap.objects) {
+    assert(
+      (object.retainedSize == null) == (object.retainer == null),
+      'retainedSize = ${object.retainedSize}, retainer = ${object.retainer}',
+    );
+    if (object.retainer != null) totalReachableSize += object.shallowSize;
+  }
+
+  assert(
+    heap.root.retainedSize == totalReachableSize,
+    '${heap.root.retainedSize} not equal to $totalReachableSize',
+  );
```

Comment:
if this should all only be done in debug mode, just wrap the entire method body in an assert